File: clients/java/cldf-java/src/main/java/io/cldf/globalid/Route.java
package io.cldf.globalid;

import lombok.Builder;

/** Represents a climbing route for CLID generation */
@Builder
public record Route(
    String name, String grade, RouteType type, FirstAscent firstAscent, Double height) {}

/** Route type enumeration */
enum RouteType {
  SPORT("sport"),
  TRAD("trad"),
  BOULDER("boulder"),
  ICE("ice"),
  MIXED("mixed");

  private final String value;

  RouteType(String value) {
    this.value = value;
  }

  @Override
  public String toString() {
    return value;
  }
}

/** First ascent information */
@Builder
record FirstAscent(Integer year, String name) {}

File: clients/dart/cldf/lib/clid/models/route.dart
/// Represents a climbing route for CLID generation
class Route {
  final String name;
  final String grade;
  final FirstAscent? firstAscent;
  final double? height;

  const Route({
    required this.name,
    required this.grade,
    this.firstAscent,
    this.height,
  });

  Map<String, dynamic> toJson() => {
    'name': name,
    'grade': grade,
    if (firstAscent != null) 'firstAscent': firstAscent!.toJson(),
    if (height != null) 'height': height,
  };

  factory Route.fromJson(Map<String, dynamic> json) {
    return Route(
      name: json['name'] as String,
      grade: json['grade'] as String,
      firstAscent: json['firstAscent'] != null
          ? FirstAscent.fromJson(json['firstAscent'] as Map<String, dynamic>)
          : null,
      height: json['height'] as double?,
    );
  }
}

/// First ascent information
class FirstAscent {
  final int? year;
  final String? name;

  const FirstAscent({this.year, this.name});

  Map<String, dynamic> toJson() => {
    if (year != null) 'year': year,
    if (name != null) 'name': name,
  };

  factory FirstAscent.fromJson(Map<String, dynamic> json) {
    return FirstAscent(
      year: json['year'] as int?,
      name: json['name'] as String?,
    );
  }
}
